Mathlib/Tactic/Linter/Header.lean

```diff
@@ -0,0 +1,293 @@
+  if (pieces.getD 1 "\n").take 1 != "\n" then
+    output := output.push (toSyntax copyright "-/", s!"{stdText "-/"}")
+  let lines := copyright.splitOn "\n"
+  let closeComment := lines.getLastD ""
+  match lines with
+  | openComment :: copyrightAuthor :: license :: authorsLines =>
+    -- The header should start and end with blank comments.
+    match openComment, closeComment with
+    | "/-", "-/" => output := output
+    | "/-", _    =>
+      output := output.push (toSyntax copyright closeComment, s!"{stdText "-/"}")
+    | _, _       =>
+      output := output.push (toSyntax copyright openComment, s!"{stdText ("/".push '-')}")
+    -- Validate the first copyright line.
+    let copStart := "Copyright (c) 20"
+    let copStop := ". All rights reserved."
+    if !copyrightAuthor.startsWith copStart then
+      output := output.push
+        (toSyntax copyright (copyrightAuthor.take copStart.length),
+         s!"Copyright line should start with 'Copyright (c) YYYY'")
+    if !copyrightAuthor.endsWith copStop then
+      output := output.push
+        (toSyntax copyright (copyrightAuthor.takeRight copStop.length),
+         s!"Copyright line should end with '. All rights reserved.'")
+    -- Validate the authors line.
+    let authorsLine := "\n".intercalate authorsLines.dropLast
+    let authorsStart := (("\n".intercalate [openComment, copyrightAuthor, license, ""])).endPos
+    output := output.append (authorsLineChecks authorsLine authorsStart)
+    let expectedLicense := "Released under Apache 2.0 license as described in the file LICENSE."
+    if license != expectedLicense then
+      output := output.push (toSyntax copyright license,
+        s!"Second copyright line should be \"{expectedLicense}\"")
+  | _ =>
+    output := output.push (toSyntax copyright "-/", s!"Copyright too short!")
+  return output
+
+/-- Check the `Syntax` `imports` for broad imports: either `Mathlib.Tactic` or any import
+starting with `Lake`. -/
+def broadImportsCheck (imports : Array Syntax)  : Array (Syntax × String) := Id.run do
+  let mut output := #[]
+  for i in imports do
+    match i.getId with
+    | `Mathlib.Tactic =>
+      output := output.push (i, s!"Files in mathlib cannot import the whole tactic folder.")
+    | modName =>
+      if modName.getRoot == `Lake then
+      output := output.push (i,
+        s!"In the past, importing 'Lake' in mathlib has led to dramatic slow-downs of the linter \
+          (see e.g. mathlib4#13779). Please consider carefully if this import is useful and \
+          make sure to benchmark it. If this is fine, feel free to allow this linter.")
+  return output
+
+/-- Check the `Syntax` imports for syntactically duplicate imports. -/
+def duplicateImportsCheck (imports : Array Syntax)  : Array (Syntax × String) := Id.run do
+  let mut output := #[]
+  let mut importsSoFar := #[]
+  for i in imports do
+    if importsSoFar.contains i then
+      output := output.push (i, s!"Duplicate imports: {i} already imported")
```

Comment:
```suggestion
      output := output.push (mkAtomFrom i s!"Duplicate imports: {i} already imported")
```
I wonder if using this, you can `return` an `Array Syntax` and then use the `Syntax` as both the position and the message when you log it.

Reply:
Good idea! I pushed a commit doing so.